bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthConnector.java

```diff
@@ -0,0 +1,341 @@
+     *
+     * @return A URL based on the authorizationEndpoint, with query parameters added.
+     * @see https://tools.ietf.org/html/rfc6749#section-4.1.1
+     */
+    public static String getAuthorizationUrl(@NonNull String authorizationEndpoint, @NonNull String clientId,
+            @Nullable String redirectURI, @Nullable String state, @Nullable String scope) {
+        StringBuilder authorizationUrl = new StringBuilder(authorizationEndpoint);
+
+        if (authorizationUrl.indexOf("?") == -1) {
+            authorizationUrl.append('?');
+        } else {
+            authorizationUrl.append('&');
+        }
+
+        try {
+            authorizationUrl.append("response_type=code");
+
+            authorizationUrl.append("&client_id=").append(URLEncoder.encode(clientId, StandardCharsets.UTF_8.name()));
+
+            if (state != null) {
+                authorizationUrl.append("&state=").append(URLEncoder.encode(state, StandardCharsets.UTF_8.name()));
+            }
+            if (redirectURI != null) {
+                authorizationUrl.append("&redirect_uri=")
+                        .append(URLEncoder.encode(redirectURI, StandardCharsets.UTF_8.name()));
+            }
+
+            if (scope != null) {
+                authorizationUrl.append("&scope=").append(URLEncoder.encode(scope, StandardCharsets.UTF_8.name()));
+            }
+
+        } catch (UnsupportedEncodingException e) {
+            // never happens
+            logger.error("Unknown encoding {}", e.getMessage(), e);
+        }
+        return authorizationUrl.toString();
+    }
+
+    /**
+     * Resource Owner Password Credentials Grant
+     *
+     * @see https://tools.ietf.org/html/rfc6749#section-4.3
+     *
+     * @param tokenUrl
+     * @param username
+     * @param password
+     * @param clientId
+     * @param clientSecret
+     * @param scope
+     * @param supportsBasicAuth
+     * @return
+     * @throws OAuthError
+     * @throws OAuthException
+     */
+    public AccessTokenResponse grantTypePassword(@NonNull String tokenUrl, @NonNull String username,
+            @NonNull String password, @Nullable String clientId, @Nullable String clientSecret, @Nullable String scope,
+            boolean supportsBasicAuth) throws OAuthError, OAuthException {
+
+        HttpClient httpClient = null;
+        try {
+            httpClient = createHttpClient(tokenUrl);
+            Request request = getMethod(httpClient, tokenUrl);
+            Fields fields = initFields(GRANT_TYPE, PASSWORD, USERNAME, username, PASSWORD, password, SCOPE, scope);
+
+            setAuthentication(clientId, clientSecret, request, fields, supportsBasicAuth);
+            return doRequest(PASSWORD, httpClient, request, fields);
+        } finally {
+            shutdownQuietly(httpClient);
+        }
+    }
+
+    /**
+     * Refresh Token
+     *
+     * @see https://tools.ietf.org/html/rfc6749#section-6
+     *
+     * @param tokenUrl
```

Comment:
if you feel param or exception descriptions would be redundant, you can remove the following lines